kube-proxy iptables - reject packets to services without endpoints

maclof opened the pull request:
When using the iptables proxy mode, reject packets to services which do not have any endpoints.

This fixes https://github.com/kubernetes/kubernetes/issues/17359

@thockin

maclof committed:
Reject packets to services without endpoints

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -459,17 +459,27 @@ func (proxier *Proxier) syncProxyRules() {
 
 	// Ensure main chains and rules are installed.
-	inputChains := []utiliptables.Chain{utiliptables.ChainOutput, utiliptables.ChainPrerouting}
+	if _, err := proxier.iptables.EnsureChain(utiliptables.TableFilter, iptablesServicesChain); err != nil {
```
I'd like to see some de-duping here.  Something like

```
tablesNeedServicesChain := []utiliptables.Table{utiliptables.TableFilter, utiliptables.TableNAT}
for _, table := tablesNeedServicesChain {
    if _, err := proxier.iptables.EnsureChain(table, iptablesServicesChain); err != nil {
        glog.Errorf("Failed to ensure that %s chain %s exists: %v", table, iptablesServicesChain, err)
        return
    }
}
```

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -459,17 +459,27 @@ func (proxier *Proxier) syncProxyRules() {
 	// Link the services chain.
+	chain := utiliptables.ChainOutput
+	comment := "kubernetes service portals"
+	args := []string{"-m", "comment", "--comment", comment, "-j", string(iptablesServicesChain)}
```
also de-dup here.  This comes down to a loop over a table of {table, chain} and all have the same body.

maclof replied:
@thockin I am only adding the rule to the OUTPUT chain on filter table right? there is no PREROUTING?

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -485,35 +495,51 @@ func (proxier *Proxier) syncProxyRules() {
 
-	chainsLines := bytes.NewBuffer(nil)
-	rulesLines := bytes.NewBuffer(nil)
+	existingNatChains := make(map[utiliptables.Chain]string)
```
s/Nat/NAT all over

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -725,42 +764,33 @@ func (proxier *Proxier) syncProxyRules() {
 	}
 
 	// Delete chains no longer in use.
-	for chain := range existingChains {
```
I think moving this actually hurts readability, any real reason for it?

maclof replied:
@thockin I did this originally because I wasn't sure which chain we'd be putting the REJECT rules onto, i'll revert it.

thockin commented:
Mostly minor points.  Thanks!

thockin commented:
ok to test

maclof commented:
I have addressed most of your feedback in my recent commit https://github.com/maclof/kubernetes/commit/d0e97939ce758f5827a143dbef3252226da142eb would like a response re the filter rules before I do the last de-dup

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -459,17 +459,25 @@ func (proxier *Proxier) syncProxyRules() {
-		args := []string{"-m", "comment", "--comment", comment, "-j", string(iptablesServicesChain)}
+	}
+	// Link the services chain.
+	chain := utiliptables.ChainOutput
```
You can dedup this as:

```
tableChainsNeedJumpServices := []struct {
    table utiliptables.Table
    chain utiliptables.Chain
} { 
    { utiliptables.TableFilter, utiliptables.ChainOutput },
    { utiliptables.TableNAT, utiliptables.ChainOutput },
    { utiliptables.TableNAT, utiliptables.ChainPrerouting },
}
comment := "kubernetes service portals"
args := []string{"-m", "comment", "--comment", comment, "-j", string(iptablesServicesChain)}
for _, tc := range tableChainsNeedJumpServices {
    if _, err := proxier.iptables.EnsureRule(utiliptables.Prepend, tc.table, tc.chain, args...); err != nil {
        glog.Errorf("Failed to ensure that %s chain %s jumps to %s: %v", tc.table, tc.chain, iptablesServicesChain, err)
        return
 }
```

thockin commented:
Last clarification made.

This also needs some e2e coverage.  Can you add case to `test/e2e/kubeproxy.go` that proves that a connection that hits this REJECT with a 5 second timeout actually takes less than 1 second (or something like that)?

thockin commented:
ok to test

thockin commented:
Any chance of this PR getting some attention in the next few days?  I'd love to get this in before the 1.2 freeze next week

maclof commented:
@thockin sorry for the delay, I've commited the dedup of the service chain rules, however I'm not very familiar with how the e2e tests work nor how to run them, but I'll take a look and see if I can make some progress with that.

thockin commented:
This change LGTM, but we really need some test for it.  I think a test for `syncProxyRules` is not out of the question, but I won't dump that on you, unless you want to be a real MVP :)

I'll LGTM this now.

thockin commented:
son of a ... something just went in that forces you to need a rebase.

thockin commented:
#18804 just landed - should be a trivial rebase.

maclof commented:
@thockin I have just rebased this, with regards to the tests, I am still learning go so not sure how much help I can be with that just now.

thockin commented:
Appreciate the contribution!

thockin commented:
Oops, I hate to be a pain - can you squash these commits?

thockin commented:
Also make sure the git email is the same one you signed the CLA with.  CLA bot is confused.

maclof commented:
@thockin should be all sorted now :)

thockin commented:
@k8s-bot test this

thockin commented:
@k8s-bot test this github issue: #20403

thockin commented:
@k8s-bot unit test this github issue: #IGNORE

thockin commented:
@k8s-bot unit test this github issue: #IGNORE

thockin commented:
@k8s-bot unit test this github issue: #IGNORE
@k8s-bot test this github issue: #IGNORE

maclof commented:
I assume these failures are flakey e2e tests?

thockin commented:
We have a static-IP leak somewhere in our test rig, which is causing tests that try to use them to fail.  Hunting it.

@k8s-bot unit test this github issue: #IGNORE
@k8s-bot test this github issue: #IGNORE

thockin commented:
@k8s-bot unit test this github issue: #20655

thockin commented:
@k8s-bot test this github issue: #IGNORE

found the right GCE project for this leak, sorry.

thockin commented:
@k8s-bot unit test this github issue: #20655

should be fixed in master now

thockin commented:
I'm going to scream. 

@maclof for sanity can you rebase to upstream/master and repush?

maclof commented:
@thockin rebased this now

thockin commented:
Hallelujah!
